---
title: "Vignette"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: default
vignette: >
  %\VignetteIndexEntry{Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, eval=FALSE}
library(Fluidigm)
```

# Preparations

## Requirements

Prior the installation of the Fluidigm package, you need to install plink, please install it according to the
Plink instructions on your system. The project webpage can be found here:

https://www.cog-genomics.org/plink2/

# Installation

## Cran (tba)
The `Fluidigm` package is hosted on Cran and can be installed via

```{r, eval=FALSE}
install.packages("Fluidigm")
```

Stable release versions are usually even numbers like 0.2, 0.4, etc. 

## GitHub
The `Fluidigm` development platform is GitHub. The main-branch is usually similar to the Cran version. Also, published release versions on
Cran are similar to Cran. Hence, the same version as Cran can be installed from GitHub

```{r, eval=FALSE}
library("devtools")
install_github("fischuu/Fluidigm")
```

Stable release versions are usually even numbers like 0.2, 0.4, etc.

However, the latest development version is hosted also on GitHub in the dev-branch and is labeled with uneven numbers 0.3.\*, 0.5.\* ,
where the \* indicates the running index of added features. These features are indicated in the CHANGELOG list and document the progress
of the package.

The latest development version of `Fluidigm` can be installed by running

```{r, eval=FALSE}
library("devtools")
install_github("fischuu/Fluidigm@dev")
```

# Usage

The package is designed to run one function after another, namely


  1. `fluidigm2PLINK(...)`
  2. `estimateErrors(...)`
  3. `calculatePairwiseSimilarities(...)`
  4. `getPairwiseSimilarityLoci(...)`
  5. `similarityMatrix(...)`
  
and the user can either run them one by one or execute them all at once, using the convenient `fluidigmAnalysisWrapper(...)`-wrapper function.

## Example data and preparations

You can download the file example_data.csv from HERE e.g. to the folder `~/My_fluidigm_project` to follow the vignette steps.

Naturally, one first needs to load the `Fluidigm` library

```{r}
library("Fluidigm")
```

For convenience, we set first the working directory to avoid playing around with file paths

```{r, eval=FALSE}
setwd("~/My_fluidigm_project")
```
## Fluidigm to Plink

The first step of the package functionality is to create a ped/map-file pair from the csv output that is received from the Fluidigm machine.

The basic use of the function is like this

```{r, eval=FALSE}
fluidigm2PLINK(file = "example_data.csv",
               map = "example_data.map",
               out = "new_data")
```

Here, the `file` parameter is the file that is received from the Fluidigm machine and the `map` option expects the location details
for the markers used in the Fluidigm, the `out` option defines the basename of the new generated files.

Other non-obligaroty input options are `out`, a string specifying the output file name. If left empty, the original basename of the input file will be used for the new generated output. Then the logical flag `plots` indicats whether additional figures for conversion should be plotted. Default is TRUE. With the logical flag `rearrange` it can be controled if the order of the created
ped/map filepair should follow the order given in the provided map file (option `TRUE`) or remain in the order as it is used in the Fluidigm file (option `FALSE`). The parameter `missing.geno` is a character string specifying how missing values should be coded. Default is "0 0". The logical option `fixNames` indicates whether whitespaces from sample names should be automatically removed. Default is TRUE. A file protection setting is the option `overwrite`. The logical indicates, wether the provided map sfile should be overwritten with the new map or not. By default, this is set to `FALSE` to avoid accidental file changes to the origianal file. If
it is set to `TRUE` without providing a new filename to `out`, the original file is overwritten, if it is set to `FALSE` the function requires an input to `out` to avoid overwriting of the original map file. As in all other functions, the functions talk-activity can be controlled with the numerical `verbosity` value. Set to a higher number for more details. Default is 1. Verbosity can be switched off with the logical flag `verbose`.

On case you want to overwrite the existing files and use the existing file names, the command would look like this

```{r, eval = FALSE}
fluidigm2PLINK(file = "example_data.csv",
               map = "example_data.map", 
               overwrite = TRUE)
```

### Output
The function generates besides the ped-file also a set of verbose plots (if the default option `plots = TRUE` is set). These are namely

1. `<filename>.additional_summary_stats.png`
2. `<filename>.call_rate_markers.png`
3. `<filename>.genotyping_success_samples.png`

The two outputs `<filename>.call_rate_markers.png` and `<filename>.genotyping_success_samples.png` show barplots to indicate the 
call rate of the genotypes and the genotyping success. The image`<filename>.additional_summary_stats.png` contains both of these,
plus some additional information like output filenames, number of samples and number of markers.

## Estimate errors

The `estimateErrors` function is a powerful tool designed to process PLINK ped files and estimate errors. It provides a comprehensive analysis of genotyping data, ensuring the accuracy and reliability of your results. This function is particularly useful in large-scale genetic studies where error estimation is crucial for data integrity.

One of the key features of `estimateErrors` is its ability to perform sex assignment and species marker analysis, if required. This is achieved by using the provided Y and X markers for sexing and species-identification markers for species analysis. 

The function is highly customizable, allowing you to specify various parameters such as the path to the ped input file, the database name, and whether new samples should be added to the database. It also allows you to control the number of replicates to keep, the markers for sexing and species identification, and the thresholds for various error checks.

Additionally, `estimateErrors` can create plots for visual inspection of the data and provides verbose output for detailed analysis. It returns a list containing a matrix indicating if genotypes are called correctly for replicates and/or if genotypes are missing, and a matrix with summary statistics.

In summary, `estimateErrors` is an essential function for anyone working with PLINK ped files, providing a robust and flexible solution for error estimation, sex assignment, and species marker analysis.

The basic use in the running example is as follows

```{r, eval=FALSE}
estErr.out <- estimateErrors(file="new_data.ped")
```

The default values to estimate the errors of the fluidigm run as `allele_error = 5`, `marker_dropout = 15` and `no_marker = 50`. That means, individuals are flags to be bad
samples, if either of the criteria is met. The allele errors are here absolte values, the marker dropout is calcualte as percentage and the number of total marker yield is again an absolute value. If the function detects bad samples, it write out a file called `<filename>.ped_samples_to_RERUN.txt` that contains samplenames from samples that
should be rerun, as they did not fulfil the minimum quality requirements.

For sexing, we need to do two things. First we need to activate the sexing module by setting `sexing=TRUE` and then add markers placed on the y-chromosome to `y.marker` and
markers from x-chromosome to `x.marker`. These two marker sets will be used for the sexing. For that marker set, the parameters provided in `male.y`, `male.hetX`, `female.y` , `female.Xtot`, `female.hetXtot`, `warning.noYtot` and `warning.noHetXtot` are used. These values give the number of detected markers, meaning, if e.g. `male.y` is set to 3, three of the `y.marker` needs to be detected in an individual to be classified to be male, then `male.hetX` gives the number of heterogeneous X-based markers need to be present and so forth.

```{r, eval=FALSE}
estErr.out <- estimateErrors(file="new_data.ped",
                             sexing = TRUE,
                             y.marker = "DBY7",
                             x.marker = c("BICF2G63", 
                                          "BICF2P19"))
```

Besides sexing, the function can also be used to flag for particular species. Here, a list of markers can be provided to the option `sp.marker` and based on that
individuals are checked for the presence of this marker set and the summary values are given in the output data matrix.

```{r, eval=FALSE}
estErr.out <- estimateErrors(file="new_data.ped",
                             sp.marker = "BICF2P5")
```

Naturally, performing sexing and species classification can be performed together like this

```{r, eval=FALSE}
estErr.out <- estimateErrors(file="new_data.ped",
                             sexing = TRUE,
                             y.marker = "DBY7",
                             x.marker = c("BICF2G63", "BICF2P19"),
                             sp.marker = "BICF2P5")
```

### Output
The function generates a couple of output files, namely

1. `<filename>.ped_samples_to_RERUN.txt`
2. `<filename>.ped_summary_individuals.csv`
3. `<filename>.new_data.GOOD.map`
4. `<filename>.new_data.GOOD.ped`
5. `<filename>.allele_error_marker_dropout.png`

Here, the first output (`<filename>.ped_samples_to_RERUN.txt`) just provides a list of samples flaged to have BAD quality. Each row contains one sample name. The file
`<filename>.ped_summary_individuals.csv` contains the same information provided as list output from the `estimateErrors()` function, see details below. The ped/map filepair
`<filename>.new_data.GOOD.map` and `<filename>.new_data.GOOD.ped` contains a filtered version of the input ped file, with bad samples removed. The map file remains unchanged
and no alleles are removed from it.

If the option `plots = TRUE` is set (default), also diagnostic plots are generated. Here, a barplot shows the frequency the different allele errors across all markers, then
another shows the marker dropout rate (in percent) across all markers and then a scatterplot of the allele error rate versus the marker droput rate.

In addition to that, the function provides the user also with a list, with two items `$gensim` and `$summs`. The `$gensim` list object provides information on the allele calls
per marker and sample. The samples are in the columns, the alleles in the rows. If an allele was called, the matrix is set to `TRUE`, if a allele is missing, the entry is set 
to `NA`. 

The `$summs` gives an informative summary data frame, depending on the picked options not all columns are always present, though. 

| Column name | Details |
|:------------|:--------|
| Ind | The name of the individual sample | 
| Ntrue | Number of alleles where replicate calls match | 
| Nfalse | Number of alleles where replicate calls do not match | 
| Nna | Number of alleles with missing data | 
| Allele_error | Number of allele errors | 
| Marker_dropout | Percentage of marker dropout | 
| No_markers_repl1 | Number of missing markers in replicate 1 | 
| No_markers_repl2 | Number of missing markers in replicate 2 | 
| categ | Overall quality assessment, based on provided thresholds | 
| noHetX1 | Number of heterogenous x-based markers for replicate 1 | 
| noHetX2 | Number of heterogenous x-based markers for replicate 2 | 
| noHetXtot | Number of total heterogenous x-based markers | 
| noX1 | Number of x-based markers for replicate 1 | 
| noX2 | Number of x-based markers for replicate 2 | 
| noXtot | Total number of x-based markers | 
| noY1 | Number of y-based markers for replicate 1 | 
| noY2 | Number of y-based markers for replicate 2 | 
| noYtot | Total number of y-based markers | 
| sex | Estiamted sex, based on provided sexing threshold | 
| noSPHetX1 | Number of heterogenous x-based species markers for replicate 1 | 
| noSPHetX2 | Number of heterogenous x-based species markers for replicate 2 | 
| noSPHetXtot | Total number of heterogenous x-based species markers| 
| noSPX1 | Number of x-based species markers for replicate 1 | 
| noSPX2 | Number of x-based species markers for replicate 2 | 
| noSPXtot | Total number of x-based species markers |
